File: SRC/CPCDOS EDITOR/Host/NameCreationService.vb
Imports System
Imports System.ComponentModel
Imports System.ComponentModel.Design.Serialization
Imports System.Diagnostics

Namespace OSMaker
    ''' <summary>
    ''' This is responsible for naming the components as they are created.
    ''' This is added as a servide by the HostSurfaceManager
    ''' </summary>
    Public Class NameCreationService
        Implements INameCreationService

        Public Sub New()
        End Sub

        Private Function CreateName(ByVal container As IContainer, ByVal type As Type) As String Implements INameCreationService.CreateName
            Dim cc = container.Components
            Dim min = Integer.MaxValue
            Dim max = Integer.MinValue
            Dim count = 0

            For i = 0 To cc.Count - 1
                Dim comp As Component = TryCast(cc(i), Component)

                If comp.GetType() Is type Then
                    count += 1
                    Dim name = comp.Site.Name

                    If name.StartsWith(type.Name) Then

                        Try
                            Dim value = Integer.Parse(name.Substring(type.Name.Length))
                            If value < min Then min = value
                            If value > max Then max = value
                        Catch ex As Exception
                            Trace.WriteLine(ex.ToString())
                        End Try
                    End If
                End If
            Next ' for
            If count = 0 Then
                Return type.Name & "1"
            ElseIf min > 1 Then
                Dim j = min - 1
                Return type.Name & j.ToString()
            Else
                Dim j = max + 1
                Return type.Name & j.ToString()
            End If
        End Function

        Private Function IsValidName(ByVal name As String) As Boolean Implements INameCreationService.IsValidName
            Return True
        End Function

        Private Sub ValidateName(ByVal name As String) Implements INameCreationService.ValidateName
            Return
        End Sub
    End Class ' class
End Namespace ' namespace

File: SRC/WindowConceptor/Host/NameCreationService.cs
using System;
using System.ComponentModel;
using System.ComponentModel.Design.Serialization;
using System.Diagnostics;

namespace WindowConceptor.Host
{
    /// <summary>
    /// This is responsible for naming the components as they are created.
    /// This is added as a servide by the HostSurfaceManager
    /// </summary>
    public class NameCreationService : INameCreationService
    {
        public NameCreationService()
        {
        }

        string INameCreationService.CreateName(IContainer container, Type type)
        {
            var cc = container.Components;
            int min = int.MaxValue;
            int max = int.MinValue;
            int count = 0;
            for (int i = 0, loopTo = cc.Count - 1; i <= loopTo; i++)
            {
                Component comp = cc[i] as Component;
                if (ReferenceEquals(comp.GetType(), type))
                {
                    count += 1;
                    string name = comp.Site.Name;
                    if (name.StartsWith(type.Name))
                    {
                        try
                        {
                            int value = int.Parse(name.Substring(type.Name.Length));
                            if (value < min)
                                min = value;
                            if (value > max)
                                max = value;
                        }
                        catch (Exception ex)
                        {
                            Trace.WriteLine(ex.ToString());
                        }
                    }
                }
            } // for

            if (count == 0)
            {
                return type.Name + "1";
            }
            else if (min > 1)
            {
                int j = min - 1;
                return type.Name + j.ToString();
            }
            else
            {
                int j = max + 1;
                return type.Name + j.ToString();
            }
        }

        bool INameCreationService.IsValidName(string name)
        {
            return true;
        }

        void INameCreationService.ValidateName(string name)
        {
            return;
        }
    } // class
} // namespace
